### ISSUE TITLE
allow smart template resolution and failover

### ISSUE THREAD

**Issue by gullerya** [gullerya/rich-component#8]
The primary concern here is to handle the case when the template content set after the initiation, yet since this is synchronous behaviour (components are initialized always eagerly) sometimes the content left empty.

Proposal:
* make eager init and if empty (specifically with the case of template) - do it again upon first consumption
* allow customization - so that API will allow to select an eager or lazy init, while for the case of **async** resolution the eager should be preferred out of the box, while for the **sync** flow - lazy.

**Comment by dankantor** [comment]
I'd like to return a function from the static getter htmlUrl() that uses properties from the instance. There is no way to do this since htmlUrl cannot receive arguments and can only return a string. 

I can do this with the getter template but I don't want to return a template, I want to return a url that is then fetched.

**Comment by gullerya** [comment]
Can you shortly describe the use case / usage. For example - would you provide a different template URL based, say, on the component instance attributes?

**Comment by dankantor** [comment]
Hi yes I would provide a different template url based on the component instance attributes. 

As an example -

`<image-gallery data-template=“inline”></image-gallery>`

Would have a different template url as 

`<image-gallery data-template=“popup”></image-gallery>`

it would even be nice if you could pick it up directly from the data attribute but I can do that in my code as well.

**Comment by gullerya** [comment]
Actually proposal to provide `data-template` with URL sounds good, I'd like to hack around with this...
> TLDR; This may require some behavioral change of the component. Today I'm injecting the HTML in c~tor, but attributes are not available yet there, so this part for this use-case will have to move to `connectedCallback`. I'm not mad on that, but need to check if still possible to provide some strong contract API that way.

One more question to be sure: so the JS part of you web component (`image-gallery`) is the same? I mean regardless of the HTML being possible different? And that is so, that it doesn't make sense for 2 separate components, each with its own constant HTML, while the whole common JS logic taken into, say, some common util and used in both?

**Comment by dankantor** [comment]
That’s correct. The JS part is the same. Just a different layout.

**Comment by gullerya** [comment]
Actually the behaviour desired here was already implemented and documented: see section of APIs `template` - `function` flavor.
PR for this issue, linked above, added a test case to exemplify the usage and enhances the documentation.

@dankantor - please notify me if you still have some issues with that pattern.

**Comment by dankantor** [comment]
@gullerya Not sure I see it. I want to dynamically return a url that is then fetched by the library. `template` only allows me to return a formed template and `htmlUrl` does not allow me to access the instance so I cannot return the dynamic url I would like. Am I missing something?

**Comment by gullerya** [comment]
Oh my, sorry... I was deeply into the dynamicity and missed that part of the ability to just provide the URL and not the template self...
Re-opening and will address it, I'll update this issue again.

**Comment by gullerya** [comment]
Please see 1.8.0, API, `htmlUrl` dynamic flavor.
Example of components that does that can be seen in this test: https://github.com/gullerya/rich-component/blob/aa080d9f74525c479740bbea93b1a22ddb7348cc/tests/test-html-url.js#L40
